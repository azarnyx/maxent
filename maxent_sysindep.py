#!/usr/bin/python
import numpy as np
from scipy.optimize import minimize
import pylab as plt
import itertools


num_sides1 = 6
num_sides2 = 10
mean1 = 4.5
mean2 = mean1


def f(x):
    if x==0:
        return 0.0
    return  - x*np.log(x)


def sum_form(p):
    s = 0
    for pi in p:
        # as we will to minimize, add minus
        s += - f(pi)
    return s


# let us consider that we have two dices. there are two possibilities:
# consider them as separate systems or consider them as one system

########### SEPARATE SYSTEMS#############
# probabilities have to be one
cons01 = ({'type': 'eq', 'fun': lambda p: np.sum(p) - 1.})
cons02 = ({'type': 'eq', 'fun': lambda p: np.sum(p) - 1.})

# introduce constraints on mean values
sides1 = np.arange(0, num_sides1)+1.
sides2 = np.arange(0, num_sides2)+1.
cons21 = ({'type': 'eq', 'fun': lambda p: sides1.dot(p) - mean1})
cons22 = ({'type': 'eq', 'fun': lambda q: sides2.dot(q) - mean2})

# bounds. do not need for entropy, however might need for other
# functions
bnds1 = tuple([(0, None)]*(num_sides1))
bnds2 = tuple([(0, None)]*(num_sides2))

# introduce initial guesses
ini_guess1 = np.array([1/num_sides1]*num_sides1)
ini_guess2 = np.array([1/num_sides2]*num_sides2)


sol1 = minimize(sum_form, ini_guess1,
                bounds=bnds1, constraints=[cons01, cons21], options={'maxiter':1001})
sol2 = minimize(sum_form, ini_guess2,
                bounds=bnds2, constraints=[cons02, cons22], options={'maxiter':1001})


########### JOINT SYSTEMS #############
# we have [p11, p12, ..., pnm] as joint pdfs
# probability>0
bnds = tuple([(0, None)]*(num_sides2*num_sides1))

# remember number of probabilities
lookup_map = {}
for ni, piqj in enumerate(itertools.product(sides1, sides2)):
    lookup_map[ni] = piqj

# common guesses
guess_common = np.ones(len(lookup_map))/len(lookup_map)

# common constrain on sum to 1
ccons0 = ({'type': 'eq', 'fun': lambda p: np.sum(p) - 1.})

# constrains on mean
def cons20_joint(params):
    s = 0
    for ni, (pij) in enumerate(params):
        s += pij*lookup_map[ni][0] # conditions for p
    return s

def cons21_joint(params):
    s = 0
    for ni, (pij) in enumerate(params):
        s += pij*lookup_map[ni][1] # conditions for q
    return s

ccons20 = ({'type': 'eq', 'fun': lambda x: cons20_joint(x)-mean1})
ccons21 = ({'type': 'eq', 'fun': lambda x: cons21_joint(x)-mean2})


sol_common = minimize(sum_form,
                      guess_common,
                      bounds=bnds,
                      constraints=[ccons0,
                                   ccons20,
                                   ccons21],
                      options={'maxiter':10001})

########### EVALUATE #############
# check p1*p2=p12
sep_sol = [i*j for i,j in  itertools.product(sol1.x, sol2.x)]
mse = np.sqrt(np.mean((sol_common.x-sep_sol)**2))
print("Difference:", mse)


# integrate on x1 and x2 to derive marginal pdf and plot them
p1_c = np.zeros(num_sides1)
p2_c = np.zeros(num_sides2)
for j, i in zip(lookup_map.items(), sol_common.x):
    p1_c[int(j[1][0])-1]+=i
    p2_c[int(j[1][1])-1]+=i

fig, ax = plt.subplots(2,1)
ax[1].bar(sides2, sol2.x, label='separate, second dice',color='C3')
ax[1].bar(sides2, p2_c, alpha=0.7, label='joint, second dice',color='C0')
ax[1].legend()
ax[0].bar(sides1, sol1.x, label='separate, first dice',color='C3')
ax[0].bar(sides1, p1_c, alpha=0.7, label='joint, first dice',color='C0')
ax[1].set_xlabel("Sides of dices")
ax[1].set_ylabel("Probaability")
ax[0].set_ylabel("Probaability")
ax[0].legend()
plt.savefig("Subsystems.png", format='png', dpi=250)
plt.show()
