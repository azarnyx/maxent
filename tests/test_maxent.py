import pytest
import sys, os
import numpy as np
sys.path.append(os.path.abspath(os.path.join(os.pardir,'')))
import maxent_sysindep, maxent_mean, maxent_variance

LI = [1e-6, 10., 1, -1, 2, -2, 0]

def test_f_sysindep(regtest):
    for i in LI:
        val = round(maxent_sysindep.f(i), 6)
        regtest.write(str(val)+' ')


def test_f_mean(regtest):
    arr = np.array(LI)
    val = maxent_mean.fun(arr)
    regtest.write(str(val)+' ')


def test_f_variance(regtest):
    arr = np.array(LI)
    val = maxent_variance.fun(arr)
    regtest.write(str(val)+' ')


def test_sum_form_sysindep(regtest):
    arr = np.array(LI)
    val = maxent_sysindep.sum_form(arr)
    regtest.write(str(val)+' ')


def test_cons_joint_sysindep(regtest):
    lookup_map = {0: (1.0, 1.0),
                  1: (1.0, 2.0),
                  2: (2.0, 1.0),
                  3: (2.0, 2.0),
                  4: (3.0, 1.0),
                  5: (3.0, 2.0)}

    p1 = [0.5, 0.5]
    p2 = [0.33333333, 0.33333333, 0.33333333]
    regtest.write(str(maxent_sysindep.cons20_joint(p1, lookup_map)))
    regtest.write(' ')
    regtest.write(str(maxent_sysindep.cons21_joint(p2, lookup_map)))
