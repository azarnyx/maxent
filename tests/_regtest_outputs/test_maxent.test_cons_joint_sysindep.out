1.0 1.33333332